DAQ/src/ArtBinaryPacketsFromDigis_module.cc

```diff
@@ -0,0 +1,1113 @@
+using  CalorimeterBoardID = mu2e::ArtFragmentReader::CalorimeterBoardID;
+using  CalorimeterHitReadoutPacket = mu2e::ArtFragmentReader::CalorimeterHitReadoutPacket;
+
+//data struct for the calorimeter
+struct CaloDataPacket {
+  CalorimeterDataPacket                    dataPacket;
+  CalorimeterBoardID                       boardID;
+  std::vector<CalorimeterHitReadoutPacket> hitPacketVec;
+  std::vector< std::vector<adc_t> >        waveformVec;
+  std::vector<uint16_t> hitIndex;
+};
+
+typedef std::deque<std::pair<mu2e_databuff_t, size_t>> raw_data_list_t;
+typedef std::pair<DataBlockHeader, TrackerDataPacket> tracker_data_block_t;
+typedef std::pair<DataBlockHeader, CaloDataPacket> calo_data_block_t;
+// typedef std::pair<DataBlockHeader,CRVDataPacket> crv_data_block_t;
+
+typedef std::deque<tracker_data_block_t> tracker_data_block_list_t;
+typedef std::deque<calo_data_block_t> calo_data_block_list_t;
+
+namespace mu2e {
+
+  constexpr int format_version = 6;
+  //--------------------------------------------------------------------
+  //
+  //
+  class ArtBinaryPacketsFromDigis : public art::EDProducer {
+  public:
+
+    explicit ArtBinaryPacketsFromDigis(fhicl::ParameterSet const& pset);
+
+    virtual void beginJob() override;
+    virtual void beginRun(art::Run&);
+
+    virtual void endJob();
+
+    void produce(art::Event&) override;
+
+    flag_mask_type filter(TrackerDataPacket& trkData,
+			  calib_constant_type clockstart,
+			  calib_constant_type panelTDCoffset, calib_constant_type hvoffset, calib_constant_type caloffset,
+			  calib_constant_type energy_max_LSHIFT8, calib_constant_type energy_min_LSHIFT8,
+			  calib_constant_type gain_RSHIFT15,
+			  calib_constant_type inverse_ionization_energy_LSHIFT26);
+
+  private:
+
+    size_t _generateTimestampTable;
+
+    // Table used for mapping between DTC timestamp and art EventID
+    std::string _tableFile;
+    std::vector< std::pair<timestamp, timestamp> > tsTable;
+
+    size_t  _timestampOffset;
+    size_t  _numWordsWritten;
+    size_t  _numEventsProcessed;
+    int     _includeTracker;
+    int     _includeCalorimeter;
+    int     _includeCosmicRayVeto;
+
+    int _includeDMAHeaders;
+
+    // Set to 1 to save packet data to a binary file
+    int _generateBinaryFile;
+
+    string                _outputFile;
+    ofstream              outputStream;
+
+    //--------------------------------------------------------------------------------
+    // TRACKER ROC/DTC INFO
+    //-------------------------------------------------------------------------------- 
+    // 96 straws per panel
+    // 1 ROC per panel
+    // 216 panels
+    //
+    // 6 ROCs per DTC
+    // 36 DTCs
+    //    const size_t number_of_rocs = 216;
+    const size_t number_of_rocs = 240;
+    const size_t number_of_straws_per_roc = 96; // Each panel in the tracker has 96 straws
+    const size_t number_of_rocs_per_dtc = 6;
+    const size_t numADCSamples = 15;
+
```

Comment:
These numbers should not be hard coded but should come from the single authoritative source.

Reply:
I completely agree. The reason why I didn't do it is that I want to do it right. I pinged the people from the sub-detectors group for having a channel map I can use for exercising with the conditions service, but I received nothing yet.